app/models/user.rb

```diff
@@ -80,8 +80,60 @@ class User < ApplicationRecord
 
   attribute :renewal_request, :string
 
-  attribute :created_at, :datetime, precision: 6
-  attribute :updated_at, :datetime, precision: 6
-
   validates :email, presence: true
+
+  # Finds, creates, or updates user from userinfo
+  # Find in case of user with existing token matching userinfo["sub"]
+  # Create in case of no token or email matching in userinfo
+  # Update in case of matching email to userinfo["email"] but no token set
+  # TODO: Add relevant security log tracking here?
+  def self.user_from_userinfo(userinfo)
+    email = userinfo[0]["email"]
+    token = userinfo[0]["sub"]
+
+    if (user = find_by(token:))
+      user
+    elsif (user = find_by(email:))
+      update_admin_added_user(user, userinfo)
+    else
+      create_user_from_userinfo(userinfo)
+    end
+  end
+
+  def self.update_admin_added_user(user, userinfo)
+    update(user.id, { token: userinfo[0]["sub"] })
+  end
+
+  def self.create_user_from_userinfo(userinfo)
+    email = userinfo[0]["email"]
+    token = userinfo[0]["sub"]
+
+    default_role_and_status = default_role_and_status_for_email(email)
+    default_role = default_role_and_status[0]
+    default_status = default_role_and_status[1]
```

Comment:
could be a one-liner
```suggestion
    default_role, default_status = default_role_and_status_for_email(email)
```